### ISSUE TITLE
cannot run on noetic or c++ 14?

### ISSUE THREAD

**Issue by yandiyang** [Livox-SDK/LIO-Livox#12]

**Comment by tranminhduc4796** [comment]
Have you solved it,  @karolmajek? I debugged LidarFeatureExtractor::FeatureExtract_with_segment and figure out that something wrong with the index in
`
data[point_num * 4 + 0] = point.x;
data[point_num * 4 + 1] = point.y;
data[point_num * 4 + 2] = point.z;
data[point_num * 4 + 3] = point.intensity;
`
Another issue that I discovered is when I printed out the value of p.x, p.y. and p.z, they are all a same constant value.

**Comment by karolmajek** [comment]
No
Tried docker with 18.04, but I had a problem with X.
Finally I'm just waiting to check it on a different laptop with 18.04, melodic (I heard it was working on such machine)

**Comment by wjh911221** [comment]
I have met the same problem

**Comment by Xu-Huitong** [comment]
Hi, friend. Have you solved this problem? I have the same problem

**Comment by ductm36** [comment]
No, but I recommend you to try out an alternative [FAST-LIO](https://github.com/hku-mars/FAST_LIO) which is also good

**Comment by zengzhiw** [comment]
this problem is caused by 
some functions do not return a value.
function : PCSeg::DoSeg  ,FreeSeg      in  segment.cpp
3 function : ComputeError   in  Estimator.h

**Comment by HowoongJun** [comment]
I had the same problem and solved exactly by what @zengzhiw said.

**Comment by FPSychotic** [comment]
> I had the same problem and solved exactly by what @zengzhiw said.
@zengzhiw
please, can you show how you do it?

**Comment by HowoongJun** [comment]
Change C++ version in CMakeLists.txt into 14
After that, do what @zengzhiw said. Set return values for those functions.

**Comment by HowoongJun** [comment]
> > I had the same problem and solved exactly by what @zengzhiw said.
> > @zengzhiw @HowoongJun
> > please, can you show how you do it?

Change C++ version in CMakeLists.txt into 14
After that, do what @zengzhiw said. Set return values for those functions.

**Comment by FPSychotic** [comment]
> > > I had the same problem and solved exactly by what @zengzhiw said.
> > > @zengzhiw @HowoongJun
> > > please, can you show how you do it?
> 
> Change C++ version in CMakeLists.txt into 14 After that, do what @zengzhiw said. Set return values for those functions.

I had fixed the cmakelist as is a quite common issue, but the return issue, I have not idea of programming, I really don't know how to do it. I asked to chatgpt , nice girl but the answer didn't work .

**Comment by HowoongJun** [comment]
> > > > I had the same problem and solved exactly by what @zengzhiw said.
> > > > @zengzhiw @HowoongJun
> > > > please, can you show how you do it?
> > 
> > 
> > Change C++ version in CMakeLists.txt into 14 After that, do what @zengzhiw said. Set return values for those functions.
> 
> I had fixed the cmakelist as is a quite common issue, but the return issue, I have not idea of programming, I really don't know how to do it. I asked to chatgpt , nice girl but the answer didn't work . Would be great if someone can copy the lines, the file or a fork for people like me.
> 
> Thanks by answer

If you carefully analyze the LIO-Livox code, you can notice that return values of those functions are actually not used. This means that any return values are okay for running this code. 
Just in case, if the function is declared in integer type (ex. int func()) the return type must integer, and if it is double, return type should be double.

**Comment by FPSychotic** [comment]
> > > > > I had the same problem and solved exactly by what @zengzhiw said.
> > > > > @zengzhiw @HowoongJun
> > > > > please, can you show how you do it?
> > > 
> > > 
> > > Change C++ version in CMakeLists.txt into 14 After that, do what @zengzhiw said. Set return values for those functions.
> > 
> > 
> > I had fixed the cmakelist as is a quite common issue, but the return issue, I have not idea of programming, I really don't know how to do it. I asked to chatgpt , nice girl but the answer didn't work . Would be great if someone can copy the lines, the file or a fork for people like me.
> > Thanks by answer
> 
> If you carefully analyze the LIO-Livox code, you can notice that return values of those functions are actually not used. This means that any return values are okay for running this code temporary. Just in case, if the function is declared in integer type (ex. int func()) the return type must integer, and if it is double, return type should be double. Therefore, this means that if the function is declared like this: int func(), possible code for return value is : return 0; int func(){ return 0; }

Thanks for your help sir , it looks compiled and it works in some way. but maybe I did not well as I cannot see any point in the topic /livox_full_cloud_mapped , does this topic works in your setup?, in RVIZ it says "Failed to transform from frame [/world] to frame [livox_frame] or [/world] depending what I use in the as global frame,(world or livox_frame).
Can anyone if that is a expected behaviour.

**Comment by HowoongJun** [comment]
> > > > > > I had the same problem and solved exactly by what @zengzhiw said.
> > > > > > @zengzhiw @HowoongJun
> > > > > > please, can you show how you do it?
> > > > 
> > > > 
> > > > Change C++ version in CMakeLists.txt into 14 After that, do what @zengzhiw said. Set return values for those functions.
> > > 
> > > 
> > > I had fixed the cmakelist as is a quite common issue, but the return issue, I have not idea of programming, I really don't know how to do it. I asked to chatgpt , nice girl but the answer didn't work . Would be great if someone can copy the lines, the file or a fork for people like me.
> > > Thanks by answer
> > 
> > 
> > If you carefully analyze the LIO-Livox code, you can notice that return values of those functions are actually not used. This means that any return values are okay for running this code temporary. Just in case, if the function is declared in integer type (ex. int func()) the return type must integer, and if it is double, return type should be double. Therefore, this means that if the function is declared like this: int func(), possible code for return value is : return 0; int func(){ return 0; }
> 
> Thanks for your help sir , it looks compiled and it works in some way. but maybe I did not well as I cannot see any point in the topic /livox_full_cloud_mapped , does this topic works in your setup?, in RVIZ it says "Failed to transform from frame [/world] to frame [livox_frame] or [/world] depending what I use in the as global frame,(world or livox_frame). Can anyone say me if that is a expected behaviour and it that topic works for you?

There is two way to solve that issue. 
First, inside rviz window "Global Options", you can set frame. Change "world" into "/world"

Second method is more fundamental solution. Inside code lio/PoseEstimation.cpp, change 513rd line "/world" into "world"

**Comment by FPSychotic** [comment]
> 

thank you sir , very kind, I will try in a while, I hope this will help others.

**Comment by FPSychotic** [comment]
> > > > > > > I had the same problem and solved exactly by what @zengzhiw said.
> > > > > > > @zengzhiw @HowoongJun
> > > > > > > please, can you show how you do it?
> > > > > 
> > > > > 
> > > > > Change C++ version in CMakeLists.txt into 14 After that, do what @zengzhiw said. Set return values for those functions.
> > > > 
> > > > 
> > > > I had fixed the cmakelist as is a quite common issue, but the return issue, I have not idea of programming, I really don't know how to do it. I asked to chatgpt , nice girl but the answer didn't work . Would be great if someone can copy the lines, the file or a fork for people like me.
> > > > Thanks by answer
> > > 
> > > 
> > > If you carefully analyze the LIO-Livox code, you can notice that return values of those functions are actually not used. This means that any return values are okay for running this code temporary. Just in case, if the function is declared in integer type (ex. int func()) the return type must integer, and if it is double, return type should be double. Therefore, this means that if the function is declared like this: int func(), possible code for return value is : return 0; int func(){ return 0; }
> > 
> > 
> > Thanks for your help sir , it looks compiled and it works in some way. but maybe I did not well as I cannot see any point in the topic /livox_full_cloud_mapped , does this topic works in your setup?, in RVIZ it says "Failed to transform from frame [/world] to frame [livox_frame] or [/world] depending what I use in the as global frame,(world or livox_frame). Can anyone say me if that is a expected behaviour and it that topic works for you?
> 
> There are two ways to solve that issue. First, inside rviz window "Global Options", you can set frame. Change "world" into "/world"
> 
> Second method is more fundamental solution. Inside code lio/PoseEstimation.cpp, change 513rd line "/world" into "world"

After tested , both solutions non worked. Anything I do , it will show world instead /world, I modified the rviz.cfg too, and removed from the launch, it always will say world instead world. In every part of any file , I checked all files and there is not more files with the word "world" I'm going to make, maybe there is missing a publisher por world no idea, Im going to try make a TF stacking publisher in the .launch /world to /livox_frame

**Comment by ixs-yuichiro-okui** [comment]
I tried to revise suspecious part, then I needed to add return in some code; then odometry and path have been updated.

https://github.com/ixs-yuichiro-okui/LIO-Livox/tree/noetic

Simply, some function in segment.cpp and estimator.h didn't own return declaration.
`estimator.h` some function should return error value but nothing in previous code. 

![image](https://github.com/Livox-SDK/LIO-Livox/assets/119822331/6ed2bda5-d9c4-4e6d-8657-a31b3945670e)


**Environment**

|Name|Value|
|---|---|
|OS|Ubuntu 20.04 with VirtualBox 6.1|
|ROS|Noetic|
|CPU Processor Number|6|

**Comment by leefan96** [comment]
我noetic 运行的特别卡，把IMU去掉才能实时运行，大佬怎么这是为什么阿
